libs/mrwebrtc-unityplugin/src/native_renderer.cpp

```diff
@@ -133,45 +156,78 @@ void NativeRenderer::DisableRemoteVideo() {
 }
 
 void NativeRenderer::I420ARemoteVideoFrameCallback(
-    void* user_data,
-    const mrsI420AVideoFrame& frame) {
+  void* user_data,
+  const mrsI420AVideoFrame& frame) {
 
-    NativeRenderer* nativeVideo = static_cast<NativeRenderer*>(user_data);
+  // It is possible for one buffer to be empty, each buffer must be checked.
+  if (frame.ydata_ == nullptr || frame.udata_ == nullptr || frame.vdata_ == nullptr) return;
 
-    // RESEARCH: Do we need to keep a frame queue or is it fine to just render
-    // the most recent frame?
+  NativeRenderer* nativeVideo = static_cast<NativeRenderer*>(user_data);
+ 
+  {
+    // Instance lock in case they get updated from UpdateTextures
+    std::lock_guard guard(nativeVideo->m_lock);
+    if ((int)frame.width_ != nativeVideo->m_remoteTextures[0].width || (int)frame.height_ != nativeVideo->m_remoteTextures[0].height) {
+      Log_Warning("NativeRenderer: I420 frame resolution changed from what it was initialized with. Resizing.");
+      g_textureSizeChangeCallback(frame.width_, frame.height_, nativeVideo->m_handle);
+      return;
+    }
+  }
 
-    // The performance trade-off being made here is to lock g_lock two times,
-    // preferring to copy the frame buffer outside any lock. Alternatively, the
-    // copy operation could be done inside g_lock scope. This would result in a
-    // single g_lock, but the lock would be held for a longer period of time
-    // every video frame, for every video stream. RESEARCH: Which is better?
+  // Acquire a video frame buffer from the free list with the global lock.
+  std::shared_ptr<I420VideoFrame> remoteI420Frame = nullptr;
+  {
+    // Global lock
+    std::lock_guard guard(g_lock);
+    if (g_freeI420VideoFrames.size()) {
+        remoteI420Frame = g_freeI420VideoFrames.back();
```

Comment:
nit: move to save 2 ref-count updates
```suggestion
        remoteI420Frame = std::move(g_freeI420VideoFrames.back());
```